src/state/auto-scroller/jump-scroller.js

```diff
@@ -0,0 +1,141 @@
+    if (!drag) {
+      console.error('Cannot move by offset when not dragging');
+      return;
+    }
+
+    const client: Position = add(drag.current.client.selection, offset);
+    move(drag.initial.descriptor.id, client, getWindowScroll(), true);
+  };
+
+  const jumpScroller: JumpScroller = (state: State) => {
+    const drag: ?DragState = state.drag;
+
+    if (!drag) {
+      return;
+    }
+
+    const request: ?Position = drag.scrollJumpRequest;
+
+    if (!request) {
+      return;
+    }
+
+    const destination: ?DraggableLocation = drag.impact.destination;
+
+    if (!destination) {
+      console.error('Cannot perform a jump scroll when there is no destination');
+      return;
+    }
+
+    const droppable: DroppableDimension = state.dimension.droppable[destination.droppableId];
+
+    // 1. We scroll the droppable first if we can to avoid the draggable
+    // leaving the list
+
+    if (canScrollDroppable(droppable, request)) {
+      const overlap: ?Position = getDroppableOverlap(droppable, request);
+
+      // Droppable can absorb the entire scroll request
+      if (!overlap) {
+        scrollDroppable(droppable.descriptor.id, request);
+        return;
+      }
+
+      // Droppable cannot absorb the entire request
+
+      // Let the droppable scroll what it can
+      const whatTheDroppableCanScroll: Position = subtract(request, overlap);
+      scrollDroppable(droppable.descriptor.id, whatTheDroppableCanScroll);
+
+      // Okay, now we need to find out where the rest of the movement can come from.
+
+      const canWindowScrollOverlap: boolean = canScrollWindow(overlap);
+
+      // window cannot absorb overlap: we need to move it
+      if (!canWindowScrollOverlap) {
+        moveByOffset(state, overlap);
+        return;
+      }
+
+      // how much can the window absorb?
+      const windowOverlap: ?Position = getWindowOverlap(overlap);
+
+      // window can absorb all of the overlap
+      if (!windowOverlap) {
+        scrollWindow(overlap);
+        return;
+      }
+
+      // window can only partially absorb overlap
+
+      const whatTheWindowCanScroll: Position = subtract(overlap, windowOverlap);
+      scrollWindow(whatTheWindowCanScroll);
+
+      // need to move the item by the remainder and scroll the window
+      moveByOffset(state, windowOverlap);
+      return;
+    }
+
+    // 2. Cannot scroll the droppable - can we scroll the window?
+
+    // Cannot scroll the window at all
+    if (!canScrollWindow(request)) {
+      moveByOffset(state, request);
+      return;
+    }
+
+    const overlap: ?Position = getWindowOverlap(request);
+
+    // Window can absorb the entire scroll
+    if (!overlap) {
+      scrollWindow(request);
+      return;
+    }
+
+    const whatTheWindowCanScroll: Position = subtract(request, overlap);
+    scrollWindow(whatTheWindowCanScroll);
+    // manually move to the rest
+    moveByOffset(state, overlap);
+  };
+
+  return jumpScroller;
+};
```

Comment:
I feel like the logic in this file is overly complicated. Can't it be simplified to something like this:

1. We have the scroll request.
2. We scroll the droppable as much as we can to fulfill the request and subtract however far it could move from the request. This gives us a remainder A.
3. We scroll the window as much as we can to fulfill remainder A and subtract that again to get remainder B.
4. We move the draggable by remainder B. Now we're done.

If the remainder is zero at the end of any step we can bail early.

I think the specific checks to see if the droppable can scroll, or absorb the entire jump, or if the window can scroll, or absorb the entire jump etc. are redundant. If the window or the droppable can't scroll then we subtract zero from the request and move to the next step. If they can absorb all of the jump then the remainder is zero so nothing will be done in the subsequent steps.

I think that's basically what you're doing, but I don't know why case 1. and case 2. in the code are being treated as different paths?

Reply:
for sure